File: frontend/scripts/generate-static-posts.js
import axios from 'axios'
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

const API_URL = process.env.VITE_API_URL || 'http://localhost:8000'

// Sample topics for generating posts
const SAMPLE_TOPICS = [
  'AI-powered productivity tools',
  'Sustainable fashion trends',
  'Tech startup innovation',
  'Digital marketing strategies',
  'Remote work culture',
  'E-commerce growth',
  'Social media engagement',
  'Content creation tips',
  'Business automation',
  'Creative design inspiration'
]

async function generateStaticPosts() {
  const postsDir = path.join(__dirname, '../public/static-posts')
  
  // Create directory if it doesn't exist
  if (!fs.existsSync(postsDir)) {
    fs.mkdirSync(postsDir, { recursive: true })
  }

  const posts = []
  
  console.log('🎨 Generating marketing posts...\n')
  
  for (let i = 0; i < SAMPLE_TOPICS.length; i++) {
    const topic = SAMPLE_TOPICS[i]
    console.log(`Generating post ${i + 1}/${SAMPLE_TOPICS.length}: ${topic}`)
    
    try {
      const response = await axios.post(
        `${API_URL}/api/marketing-post/create`,
        {
          topic,
          caption_style: 'engaging',
          aspect_ratio: '1:1',
          include_hashtags: true,
          post_to_instagram: false
        },
        { timeout: 60000 } // 60 second timeout
      )
      
      const postData = response.data
      
      // Save image
      let imagePath = null
      if (postData.image_base64) {
        const imageBuffer = Buffer.from(postData.image_base64, 'base64')
        imagePath = `static-posts/post-${i + 1}.png`
        const fullImagePath = path.join(postsDir, `post-${i + 1}.png`)
        fs.writeFileSync(fullImagePath, imageBuffer)
        console.log(`  ✓ Image saved: ${imagePath}`)
      } else if (postData.image_url) {
        // Download image from URL
        const imageResponse = await axios.get(postData.image_url, { responseType: 'arraybuffer' })
        imagePath = `static-posts/post-${i + 1}.png`
        const fullImagePath = path.join(postsDir, `post-${i + 1}.png`)
        fs.writeFileSync(fullImagePath, imageResponse.data)
        console.log(`  ✓ Image saved: ${imagePath}`)
      }
      
      // Create post metadata
      const postMetadata = {
        id: i,
        topic,
        caption: postData.full_caption || postData.caption,
        image: imagePath,
        hashtags: postData.hashtags || [],
        image_prompt: postData.image_prompt || null
      }
      
      posts.push(postMetadata)
      console.log(`  ✓ Post ${i + 1} complete\n`)
      
      // Small delay to avoid rate limiting
      await new Promise(resolve => setTimeout(resolve, 1000))
      
    } catch (error) {
      console.error(`  ✗ Error generating post ${i + 1}:`, error.message)
      // Create a placeholder post
      const postMetadata = {
        id: i,
        topic,
        caption: `Engaging content about ${topic}. Discover how this topic can transform your business and drive results.`,
        image: null, // Will use gradient
        hashtags: [],
        image_prompt: null
      }
      posts.push(postMetadata)
    }
  }
  
  // Save posts metadata as JSON
  const metadataPath = path.join(postsDir, 'posts.json')
  fs.writeFileSync(metadataPath, JSON.stringify(posts, null, 2))
  console.log(`\n✅ All posts generated!`)
  console.log(`📁 Posts saved to: ${postsDir}`)
  console.log(`📄 Metadata saved to: ${metadataPath}`)
  console.log(`\nTotal posts: ${posts.length}`)
}

generateStaticPosts().catch(console.error)

File: generate_static_posts.py
import os
import requests
import json
import base64
import time
from pathlib import Path

API_URL = "http://localhost:8000"

# Sample topics for generating posts
SAMPLE_TOPICS = [
    'AI-powered productivity tools',
    'Sustainable fashion trends',
    'Tech startup innovation',
    'Digital marketing strategies',
    'Remote work culture',
    'E-commerce growth',
    'Social media engagement',
    'Content creation tips',
    'Business automation',
    'Creative design inspiration'
]

def wait_for_backend(max_retries=30, delay=2):
    """Wait for backend to be available"""
    print("⏳ Waiting for backend server to be ready...")
    for i in range(max_retries):
        try:
            response = requests.get(f"{API_URL}/api/health", timeout=2)
            if response.status_code == 200:
                print("✅ Backend server is ready!\n")
                return True
        except:
            pass
        if i < max_retries - 1:
            print(f"   Attempt {i + 1}/{max_retries}...")
            time.sleep(delay)
    print("❌ Backend server not available. Please start the backend first.")
    return False

def generate_static_posts():
    # Create directory for static posts
    posts_dir = Path("frontend/public/static-posts")
    posts_dir.mkdir(parents=True, exist_ok=True)
    
    posts = []
    
    print("🎨 Generating marketing posts...\n")
    
    for i, topic in enumerate(SAMPLE_TOPICS):
        print(f"Generating post {i + 1}/{len(SAMPLE_TOPICS)}: {topic}")
        
        try:
            response = requests.post(
                f"{API_URL}/api/marketing-post/create",
                json={
                    "topic": topic,
                    "caption_style": "engaging",
                    "aspect_ratio": "1:1",
                    "include_hashtags": True,
                    "post_to_instagram": False
                },
                timeout=60
            )
            
            if response.status_code != 200:
                print(f"  ✗ Error: {response.status_code} - {response.text}")
                # Create placeholder
                posts.append({
                    "id": i,
                    "topic": topic,
                    "caption": f"Engaging content about {topic}. Discover how this topic can transform your business and drive results.",
                    "image": None,
                    "hashtags": [],
                    "image_prompt": None
                })
                continue
            
            post_data = response.json()
            
            # Save image
            image_path = None
            if post_data.get("image_base64"):
                image_buffer = base64.b64decode(post_data["image_base64"])
                image_filename = f"post-{i + 1}.png"
                image_path = f"static-posts/{image_filename}"
                full_image_path = posts_dir / image_filename
                full_image_path.write_bytes(image_buffer)
                print(f"  ✓ Image saved: {image_path}")
            elif post_data.get("image_url"):
                # Download image from URL
                img_response = requests.get(post_data["image_url"], timeout=30)
                if img_response.status_code == 200:
                    image_filename = f"post-{i + 1}.png"
                    image_path = f"static-posts/{image_filename}"
                    full_image_path = posts_dir / image_filename
                    full_image_path.write_bytes(img_response.content)
                    print(f"  ✓ Image saved: {image_path}")
            
            # Create post metadata
            post_metadata = {
                "id": i,
                "topic": topic,
                "caption": post_data.get("full_caption") or post_data.get("caption", ""),
                "image": image_path,
                "hashtags": post_data.get("hashtags", []),
                "image_prompt": post_data.get("image_prompt")
            }
            
            posts.append(post_metadata)
            print(f"\n✅ Post {i + 1} complete!")
            print(f"   Topic: {topic}")
            print(f"   Image: {image_path or 'Gradient placeholder'}")
            print(f"   Caption length: {len(post_metadata['caption'])} chars")
            
            # Wait before next post to avoid rate limiting
            if i < len(SAMPLE_TOPICS) - 1:
                print(f"\n⏳ Waiting 3 seconds before next post...")
                time.sleep(3)
            
        except Exception as error:
            print(f"\n❌ Error generating post {i + 1}: {str(error)}")
            # Create a placeholder post
            posts.append({
                "id": i,
                "topic": topic,
                "caption": f"Engaging content about {topic}. Discover how this topic can transform your business and drive results.",
                "image": None,
                "hashtags": [],
                "image_prompt": None
            })
            print(f"   Using placeholder for post {i + 1}")
            
            # Wait before next post even on error
            if i < len(SAMPLE_TOPICS) - 1:
                print(f"\n⏳ Waiting 2 seconds before next post...")
                time.sleep(2)
    
    # Save posts metadata as JSON
    metadata_path = posts_dir / "posts.json"
    with open(metadata_path, "w") as f:
        json.dump(posts, f, indent=2)
    
    print(f"\n{'='*60}")
    print(f"✅ All posts generated!")
    print(f"{'='*60}")
    print(f"📁 Posts saved to: {posts_dir}")
    print(f"📄 Metadata saved to: {metadata_path}")
    print(f"\nTotal posts: {len(posts)}")
    print(f"Posts with images: {sum(1 for p in posts if p.get('image'))}")
    print(f"Placeholder posts: {sum(1 for p in posts if not p.get('image'))}")

if __name__ == "__main__":
    generate_static_posts()

